prov/rxm/src/rxm_ep.c

```diff
@@ -1056,84 +1013,134 @@ rxm_ep_sar_tx_send(struct rxm_ep *rxm_ep, struct rxm_conn *rxm_conn, void *conte
+	return ret;
+}
+
+static inline ssize_t
+rxm_ep_send_deferred_tx(struct rxm_ep *rxm_ep, struct rxm_conn *rxm_conn,
+			struct rxm_tx_entry *def_tx_entry)
 {
 	ssize_t ret;
-	size_t tx_size = sizeof(struct rxm_pkt) + tx_entry->tx_buf->pkt.hdr.size;
+	struct rxm_tx_entry *tx_entry = rxm_tx_entry_get(&rxm_conn->send_queue);
+	if (OFI_UNLIKELY(!tx_entry))
+		return -FI_EAGAIN;
 
-	tx_entry->tx_buf->pkt.ctrl_hdr.conn_id = rxm_conn->handle.remote_key;
-	FI_DBG(&rxm_prov, FI_LOG_EP_DATA,
-	       "Send deferred TX request (len - %zd) for %p conn\n",
-	       tx_entry->tx_buf->pkt.hdr.size, rxm_conn);
-
-	if ((tx_size <= rxm_ep->msg_info->tx_attr->inject_size) &&
-	    (tx_entry->flags & FI_INJECT) && !(tx_entry->flags & FI_COMPLETION))  {
-		(void) rxm_ep_inject_send(rxm_ep, rxm_conn,
-					  tx_entry->tx_buf, tx_size);
-		/* Release TX entry for futher reuse */
-		rxm_tx_entry_release(&rxm_conn->send_queue, tx_entry);
-	} else if (tx_entry->tx_buf->pkt.hdr.size >
-			rxm_ep->rxm_info->tx_attr->inject_size) {
-		struct rxm_rma_iov *rma_iov =
-			(struct rxm_rma_iov *)&tx_entry->tx_buf->pkt.data;
-		ret = rxm_ep_lmt_tx_send(rxm_ep, rxm_conn, tx_entry,
-					 sizeof(struct rxm_pkt) + sizeof(*rma_iov) +
-					 sizeof(*rma_iov->iov) * tx_entry->count);
-		if (OFI_UNLIKELY(ret)) {
-			FI_WARN(&rxm_prov, FI_LOG_EP_DATA,
-				"Unable to perform deferred large send operation\n");
-			rxm_cq_write_error(rxm_ep->util_ep.tx_cq,
-					   rxm_ep->util_ep.tx_cntr,
-					   tx_entry->context, (int)ret);
-		}
+	tx_entry->state = def_tx_entry->state;
+	rxm_fill_tx_entry(def_tx_entry->context, def_tx_entry->count,
+			  def_tx_entry->flags, def_tx_entry->comp_flags,
+			  def_tx_entry->tx_buf, tx_entry);
+
+	ret = rxm_ep_normal_send(rxm_ep, rxm_conn, tx_entry,
+				 tx_entry->tx_buf->pkt.hdr.size +
+				 sizeof(struct rxm_pkt));
+	if (!ret) {
+		dlist_remove(&def_tx_entry->deferred_entry);
+		free(def_tx_entry);
 	} else {
-		ret = rxm_ep_normal_send(rxm_ep, rxm_conn, tx_entry, tx_size);
-		if (OFI_UNLIKELY(ret)) {
-			FI_WARN(&rxm_prov, FI_LOG_EP_DATA,
-				"Unable to perform deferred send operation\n");
-			rxm_cq_write_error(rxm_ep->util_ep.tx_cq,
-					   rxm_ep->util_ep.tx_cntr,
-					   tx_entry->context, (int)ret);
+		rxm_tx_entry_release(&rxm_conn->send_queue, tx_entry);
+	}
+	return ret;
+}
+
+void rxm_ep_progress_conn_deferred_list(struct rxm_ep *rxm_ep, struct rxm_conn *rxm_conn)
+{
+	ssize_t ret = 0;
+
+	while (!dlist_empty(&rxm_conn->deferred_op_list) && !ret) {
+		struct rxm_tx_entry *tx_entry =
+			container_of(rxm_conn->deferred_op_list.next,
+				     struct rxm_tx_entry, deferred_entry);
+		if ((tx_entry->tx_buf->pkt.hdr.size + sizeof(struct rxm_pkt)) <=
+				rxm_ep->msg_info->tx_attr->inject_size) {
+			ret = rxm_ep_inject_deferred_tx(rxm_ep, rxm_conn, tx_entry);
+		} else {
+			ret = rxm_ep_send_deferred_tx(rxm_ep, rxm_conn, tx_entry);
 		}
 	}
+
+	if (dlist_empty(&rxm_conn->deferred_op_list))
+		dlist_remove_init(&rxm_conn->conn_deferred_entry);
```

Comment:
nit: I think it would be cleaner to move this into rxm_ep_progress_deferred_list() loop.

Reply:
Sure, I fixed this